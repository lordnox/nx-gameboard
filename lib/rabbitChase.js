import * as Seedrandom from 'seedrandom';
const seedrandom = Seedrandom;
import { drawBoard, generateGameboard } from './gameboard';
import { Solution } from './solution';
const MAX_DISTANCE = 999999;
const empty = ' ';
const rabbit = '🐇';
const carrot = '🥕';
const initializeSeed = (seed) => `${seed ? seed : seedrandom(`${Math.random()}`, { entropy: true })()}`;
export const RabbitChaseBoard = ({ seed, cols = 6, rows = 6, maxItems = Math.floor((cols / 2) + (rows / 2) + 1), } = {}) => {
    const usedSeed = initializeSeed(seed);
    const rng = seedrandom(usedSeed);
    const settings = {
        cols,
        item: {
            hidden: rabbit,
            clue: carrot,
        },
        maxItems,
        rows,
        seed: usedSeed,
    };
    const board = generateGameboard({
        cols: settings.cols,
        rows: settings.rows,
        carrots: settings.maxItems,
        cell: (baseCell) => (Object.assign({}, baseCell, { distance: MAX_DISTANCE, state: empty, draw(showHidden) {
                if (this.state === empty || !showHidden && this.state === settings.item.hidden)
                    return '   ';
                return ` ${this.state}`;
            } })),
    });
    const getInt = (max) => Math.floor(rng() * max);
    const getIndex = (list) => getInt(list.length);
    const getItem = (list) => list[getIndex(list)];
    const clear = () => {
        board.cells().forEach(cell => cell.state = empty);
    };
    const findByType = (state) => board.cells().filter(cell => cell.state === state);
    const findNeighbours = ({ x, y }) => [
        board.get(x - 1, y - 1), board.get(x, y - 1), board.get(x + 1, y - 1),
        board.get(x - 1, y), board.get(x + 1, y),
        board.get(x - 1, y + 1), board.get(x, y + 1), board.get(x + 1, y + 1),
    ].filter(x => x);
    const removeCellFromList = (list) => (cell) => {
        const index = list.findIndex(item => item === cell);
        return index === -1 ? null : list.splice(index, 1);
    };
    const findDirectNeighbours = ({ x, y }) => [
        board.get(x, y - 1), board.get(x - 1, y), board.get(x + 1, y), board.get(x, y + 1),
    ].filter(x => x);
    const init = () => {
        clear();
        const availableCells = board.cells().map(cell => cell.index);
        const remove = removeCellFromList(availableCells);
        const setHidden = (index) => {
            const cell = board.cell(index);
            cell.state = settings.item.hidden;
            remove(index);
            cell.distance = 0;
            const neighbours = findNeighbours(cell).map(cell => cell.index);
            neighbours.map(remove);
            neighbours.forEach(index => board.cell(index).distance = '%');
            availableCells.forEach(index => {
                const item = board.cell(index);
                const dX = Math.abs(item.x - cell.x);
                const dY = Math.abs(item.y - cell.y);
                const d = Math.floor(Math.sqrt(dX * dX + dY * dY));
                item.distance = d;
            });
        };
        const findAndSetCell = (cells) => {
            setHidden(getItem(cells));
            if (!availableCells.length)
                return [];
            const lowDistance = availableCells.reduce((memo, index) => {
                const { distance } = board.cell(index);
                return distance === '%' || memo < distance ? memo : distance;
            }, MAX_DISTANCE);
            return availableCells.filter(index => board.cell(index).distance <= lowDistance + 1);
        };
        // add as many carrots as possible
        let cells = availableCells;
        while (cells.length)
            cells = findAndSetCell(cells);
        const hiddenItems = board.cells().filter(cell => cell.state === settings.item.hidden);
        while (hiddenItems.length > settings.maxItems) {
            const index = getIndex(hiddenItems);
            hiddenItems[index].state = empty;
            hiddenItems.splice(index, 1);
        }
        hiddenItems.forEach(item => {
            const possibleRabbits = findDirectNeighbours(item)
                .filter(cell => cell.state === empty);
            getItem(possibleRabbits).state = settings.item.clue;
        });
    };
    const collectFromCells = (state) => (cells) => cells.reduce((memo, cellGetter) => memo + (cellGetter().state === state ? 1 : 0), 0);
    const toJSON = () => {
        const top = board.cols.map(collectFromCells(settings.item.hidden));
        const left = board.rows.map(collectFromCells(settings.item.hidden));
        const carrots = findByType(settings.item.clue).map(({ x, y, index }) => ({ x, y, index }));
        return {
            cols: settings.cols,
            rows: settings.rows,
            carrots,
            top,
            left,
        };
    };
    const generateOutputCellState = ({ left, top }, allowHidden) => ({ x, y }) => {
        if (x === 0) {
            if (y === 0)
                return empty;
            return left[y - 1];
        }
        if (y === 0)
            return top[x - 1];
        const cell = board.get(x - 1, y - 1);
        if (!cell)
            return empty;
        const { state } = cell;
        if (allowHidden)
            return state;
        return state === settings.item.hidden ? empty : state;
    };
    return {
        init,
        drawBoard: (showHidden = false) => {
            const json = toJSON();
            const cellState = generateOutputCellState(json, showHidden);
            const localBoard = generateGameboard({
                cols: settings.cols + 1,
                rows: settings.rows + 1,
                cell: cell => {
                    return Object.assign({}, cell, { state: cellState(cell), draw(showHidden) {
                            if (typeof this.state === 'number')
                                return ` ${this.state} `;
                            if (this.state === empty)
                                return '   ';
                            return ` ${this.state}`;
                        } });
                },
            });
            const emptyCell = { draw: () => '   ' };
            return drawBoard(localBoard, showHidden);
        },
        getInitialSolution: () => {
            const hidden = [empty, rabbit];
            const unknown = '?';
            const actions = ['cellToEmpty', 'cellToUnknown', 'cellToRabbit'];
            const sumHints = (cells) => cells.reduce((memo, cell) => memo + (cell.state === rabbit ? 1 : 0), 0);
            return Solution(board, cell => hidden.includes(cell.state), unknown, cell => cell.state, sumHints, actions, (cells, action, payload) => cells);
        },
        toJSON,
    };
};
//# sourceMappingURL=rabbitChase.js.map