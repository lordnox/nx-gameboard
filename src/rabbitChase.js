"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
exports.__esModule = true;
var seedrandom = require("seedrandom");
var index_1 = require("./index");
var MAX_DISTANCE = 999999;
var empty = ' ';
var rabbit = '🐇';
var carrot = '🥕';
exports.RabbitChase = function (_a) {
    var _b = _a.seed, seed = _b === void 0 ? 'RabbitChase' : _b;
    var rng = seedrandom(seed);
    var cols = 5;
    var rows = 5;
    var maxCarrots = 4;
    var board = index_1.generateGameboard({
        seed: seed, cols: cols, rows: rows, carrots: maxCarrots,
        cell: function (baseCell) { return (__assign({}, baseCell, { distance: MAX_DISTANCE, state: empty, draw: function () { return " " + this.state + " "; } })); }
    });
    var getInt = function (max) { return Math.floor(rng() * max); };
    var getIndex = function (list) { return getInt(list.length); };
    var getItem = function (list) { return list[getIndex(list)]; };
    var check = function () { return null; };
    var place = function (x, y) {
        var cell = board.get(x, y);
    };
    var clear = function () {
        board.cells().forEach(function (cell) { return cell.state = empty; });
    };
    var findNeighbours = function (_a) {
        var x = _a.x, y = _a.y;
        return [
            board.get(x - 1, y - 1), board.get(x, y - 1), board.get(x + 1, y - 1),
            board.get(x - 1, y), board.get(x + 1, y),
            board.get(x - 1, y + 1), board.get(x, y + 1), board.get(x + 1, y + 1),
        ].filter(function (x) { return x; });
    };
    var removeCellFromList = function (list) { return function (cell) {
        var index = list.findIndex(function (item) { return item === cell; });
        if (index === -1)
            return null;
        return list.splice(index, 1);
    }; };
    var findDirectNeighbours = function (_a) {
        var x = _a.x, y = _a.y;
        return [
            board.get(x, y - 1), board.get(x - 1, y), board.get(x + 1, y), board.get(x, y + 1),
        ].filter(function (x) { return x; });
    };
    var init = function () {
        clear();
        var availableCells = board.cells().map(function (cell) { return cell.index; });
        var remove = removeCellFromList(availableCells);
        var setCarrot = function (index, d) {
            if (d === void 0) { d = false; }
            var cell = board.cell(index);
            cell.state = carrot;
            remove(index);
            cell.distance = 0;
            var neighbours = findNeighbours(cell).map(function (cell) { return cell.index; });
            if (d)
                console.log(neighbours);
            neighbours.map(remove);
            neighbours.forEach(function (index) { return board.cell(index).distance = '%'; });
            availableCells.forEach(function (index) {
                var item = board.cell(index);
                var dX = Math.abs(item.x - cell.x);
                var dY = Math.abs(item.y - cell.y);
                var d = Math.floor(Math.sqrt(dX * dX + dY * dY));
                item.distance = d;
            });
        };
        var findAndSetCell = function (cells) {
            setCarrot(getItem(cells));
            if (!availableCells.length)
                return [];
            var lowDistance = availableCells.reduce(function (memo, index) {
                var distance = board.cell(index).distance;
                return memo < distance ? memo : distance;
            }, MAX_DISTANCE);
            return availableCells.filter(function (index) { return board.cell(index).distance <= lowDistance + 1; });
        };
        // add as many carrots as possible
        var cells = availableCells;
        while (cells.length)
            cells = findAndSetCell(cells);
        var carrots = board.cells().filter(function (cell) { return cell.state === carrot; });
        while (carrots.length > maxCarrots) {
            var index = getIndex(carrots);
            carrots[index].state = empty;
            carrots.splice(index, 1);
        }
        carrots.forEach(function (carrot) {
            var possibleRabbits = findDirectNeighbours(carrot)
                .filter(function (cell) { return cell.state === empty; });
            getItem(possibleRabbits).state = rabbit;
        });
        // const rabbitPlaces = arrayUnique(arrayConcat()
        // rabbitPlaces.forEach(cell => cell.state = rabbit)
    };
    return { init: init, place: place, drawBoard: function () { return index_1.drawBoard(board); } };
};
