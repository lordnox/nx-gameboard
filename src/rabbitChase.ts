
import { default as seedrandom } from 'seedrandom'

import { arrayConcat, arrayUnique, BaseCell, drawBoard, generateGameboard } from './gameboard'

const MAX_DISTANCE = 999999
const empty = ' '
const rabbit = '🐇'
const carrot = '🥕'

export type CellState = typeof empty | typeof carrot | typeof rabbit
export type RabbitChaseCellGetter = () => RabbitChaseCell
export type Distance = number | '%'
export interface RabbitChaseCell extends BaseCell {
  state: CellState
  distance: Distance
}
export interface RabbitChaseOutputCell extends BaseCell {
  state: CellState | number
}
export interface RabbitChaseOptions {
  seed: string | number,
  cols: number
  rows: number
  maxItems: number
}

const initializeSeed = (seed?: string | number) =>
  `${seed ? seed : seedrandom(`${Math.random()}`, { entropy: true })()}`

export interface RabbitChaseBoardSettingsItem {
  hidden: CellState
  clue: CellState
}

export interface RabbitChaseBoardSettings extends RabbitChaseOptions {
  maxItems: number
  item: RabbitChaseBoardSettingsItem
}

export const RabbitChaseBoard = ({
  seed,
  cols = 6,
  rows = 6,
  maxItems = Math.floor((cols / 2) + (rows / 2) + 1),
}: Partial<RabbitChaseOptions> = {}) => {
  const usedSeed = initializeSeed(seed)
  const rng = seedrandom(usedSeed)

  const settings: RabbitChaseBoardSettings = {
    cols,
    item: {
      hidden: rabbit,
      clue: carrot,
    },
    maxItems,
    rows,
    seed: usedSeed,
  }

  const board = generateGameboard<RabbitChaseCell>({
    cols: settings.cols,
    rows: settings.rows,
    carrots: settings.maxItems,
    cell: (baseCell): RabbitChaseCell => ({
      ...baseCell,
      distance: MAX_DISTANCE,
      state: empty,
      draw(showHidden) {
        if(this.state === empty || !showHidden && this.state === settings.item.hidden)
          return '   '
        return ` ${this.state}`
      },
    }),
  })

  const getInt = (max: number) => Math.floor(rng() * max)
  const getIndex = (list: any[]) => getInt(list.length)
  const getItem = <Item>(list: Item[]) => list[getIndex(list)]

  const clear = () => {
    board.cells().forEach(cell => cell.state = empty)
  }

  const findByType = (state: CellState) => board.cells().filter(cell => cell.state === state)

  const findNeighbours = ({ x, y }: BaseCell) => [
    board.get(x - 1, y - 1), board.get(x, y - 1), board.get(x + 1, y - 1),
    board.get(x - 1, y),                          board.get(x + 1, y),
    board.get(x - 1, y + 1), board.get(x, y + 1), board.get(x + 1, y + 1),
  ].filter(x => x)

  const removeCellFromList = (list: number[]) => (cell: number) => {
    const index = list.findIndex(item => item === cell)
    return index === -1 ? null : list.splice(index, 1)
  }

  const findDirectNeighbours = ({ x, y }: BaseCell) => [
    board.get(x, y - 1), board.get(x - 1, y), board.get(x + 1, y), board.get(x, y + 1),
  ].filter(x => x) as RabbitChaseCell[]

  const init = () => {
    clear()
    const availableCells = board.cells().map(cell => cell.index)
    const remove = removeCellFromList(availableCells)

    const setHidden = (index: number) => {
      const cell = board.cell(index)
      cell.state = settings.item.hidden
      remove(index)
      cell.distance = 0
      const neighbours = findNeighbours(cell).map(cell => cell.index)
      neighbours.map(remove)
      neighbours.forEach(index => board.cell(index).distance = '%')
      availableCells.forEach(index => {
        const item = board.cell(index)
        const dX = Math.abs(item.x - cell.x)
        const dY = Math.abs(item.y - cell.y)
        const d = Math.floor(Math.sqrt(dX * dX + dY * dY))
        item.distance = d
      })
    }

    const findAndSetCell = (cells: number[]) => {
      setHidden(getItem(cells))
      if(!availableCells.length)
        return []
      const lowDistance = availableCells.reduce((memo, index) => {
        const { distance } = board.cell(index)
        return distance === '%' || memo < distance ? memo : distance
      }, MAX_DISTANCE)
      return availableCells.filter(index => board.cell(index).distance <= lowDistance + 1)
    }

    // add as many carrots as possible
    let cells = availableCells
    while(cells.length)
      cells = findAndSetCell(cells)

    const hiddenItems = board.cells().filter(cell => cell.state === settings.item.hidden)
    while(hiddenItems.length > settings.maxItems) {
      const index = getIndex(hiddenItems)
      hiddenItems[index].state = empty
      hiddenItems.splice(index, 1)
    }

    hiddenItems.forEach(item => {
      const possibleRabbits = findDirectNeighbours(item)
      .filter(cell => cell.state === empty)
      getItem(possibleRabbits).state = settings.item.clue
    })
  }

  const collectFromCells = (state: CellState) => (cells: RabbitChaseCellGetter[]) =>
    cells.reduce((memo, cellGetter) => memo + (cellGetter().state === state ? 1 : 0), 0)

  const toJSON = () => {
    const top = board.cols.map(collectFromCells(settings.item.hidden))
    const left = board.rows.map(collectFromCells(settings.item.hidden))
    const carrots = findByType(settings.item.clue).map(({ x, y, index }) => ({ x, y, index }))
    return {
      cols: settings.cols,
      rows: settings.rows,
      carrots,
      top,
      left,
    }
  }

  const generateOutputCellState = ({ left, top }: ReturnType<typeof toJSON>, allowHidden: boolean) =>
    ({ x, y }: BaseCell) => {
      if(x === 0) {
        if(y === 0)
          return empty
        return left[y - 1]
      }
      if(y === 0)
        return top[x - 1]
      const cell = board.get(x - 1, y - 1)
      if(!cell)
        return empty
      const { state } = cell
      if(allowHidden)
        return state
      return state === settings.item.hidden ? empty : state
    }

  return {
    init,
    drawBoard: (showHidden = false) => {
      const json = toJSON()
      const cellState = generateOutputCellState(json, showHidden)
      const localBoard = generateGameboard<RabbitChaseOutputCell>({
        cols: settings.cols + 1,
        rows: settings.rows + 1,
        cell: cell => {
          return {
            ...cell,
            state: cellState(cell),
            draw(showHidden) {
              if(typeof this.state === 'number')
                return ` ${this.state} `
              if(this.state === empty)
                return '   '
              return ` ${this.state}`
            },
          }
        },
      })
      const emptyCell = { draw: () => '   ' }
      return drawBoard(localBoard, showHidden)
    },
    getInitialSolution: () => {
      console.log('solution')
    },
    toJSON,
  }
}
