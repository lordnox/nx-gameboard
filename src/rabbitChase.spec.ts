
import { RabbitChaseBoard } from './rabbitChase'

describe('rabbitChase', () => {
  it('should generate a game board', () => {
    const rabbitChase = RabbitChaseBoard({ seed: 'test-a' })
    rabbitChase.init()
    expect(rabbitChase.drawBoard()).toMatchSnapshot()

    const rabbitChase1 = RabbitChaseBoard({ seed: 'test-b' })
    rabbitChase.init()
    expect(rabbitChase.drawBoard()).toMatchSnapshot()

    const rabbitChase2 = RabbitChaseBoard({ seed: 'test-c' })
    rabbitChase.init()
    expect(rabbitChase.drawBoard()).toMatchSnapshot()
  })

  it('should render it to json correctly', () => {
    const rabbitChase = RabbitChaseBoard({ seed: 'test' })
    rabbitChase.init()
    rabbitChase.getInitialSolution()
    expect(rabbitChase.toJSON()).toMatchSnapshot()
    expect(rabbitChase.drawBoard()).toMatchSnapshot()
    expect(rabbitChase.drawBoard(true)).toMatchSnapshot()
  })

  it('should be able to change the settings', () => {
    const rabbitChase66 = RabbitChaseBoard({ seed: 'test' })
    const rabbitChase88 = RabbitChaseBoard({ seed: 'test', cols: 8, rows: 8, maxItems: 50 })
    rabbitChase66.init()
    rabbitChase88.init()
    expect(rabbitChase88.toJSON()).not.toEqual(rabbitChase66.toJSON())
    expect(rabbitChase88.toJSON()).toMatchSnapshot()
    expect(rabbitChase88.drawBoard()).toMatchSnapshot()
    expect(rabbitChase88.drawBoard(true)).toMatchSnapshot()
  })
})
