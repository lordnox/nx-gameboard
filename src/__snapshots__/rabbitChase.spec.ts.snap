// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rabbitChase should be able to change the settings 1`] = `
Object {
  "carrots": Array [
    Object {
      "index": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "index": 6,
      "x": 6,
      "y": 0,
    },
    Object {
      "index": 11,
      "x": 3,
      "y": 1,
    },
    Object {
      "index": 12,
      "x": 4,
      "y": 1,
    },
    Object {
      "index": 21,
      "x": 5,
      "y": 2,
    },
    Object {
      "index": 22,
      "x": 6,
      "y": 2,
    },
    Object {
      "index": 32,
      "x": 0,
      "y": 4,
    },
    Object {
      "index": 33,
      "x": 1,
      "y": 4,
    },
    Object {
      "index": 37,
      "x": 5,
      "y": 4,
    },
    Object {
      "index": 43,
      "x": 3,
      "y": 5,
    },
    Object {
      "index": 47,
      "x": 7,
      "y": 5,
    },
    Object {
      "index": 51,
      "x": 3,
      "y": 6,
    },
    Object {
      "index": 52,
      "x": 4,
      "y": 6,
    },
    Object {
      "index": 56,
      "x": 0,
      "y": 7,
    },
    Object {
      "index": 62,
      "x": 6,
      "y": 7,
    },
  ],
  "cols": 8,
  "left": Array [
    2,
    2,
    2,
    1,
    1,
    3,
    1,
    3,
  ],
  "rows": 8,
  "top": Array [
    2,
    2,
    2,
    1,
    3,
    1,
    1,
    3,
  ],
}
`;

exports[`rabbitChase should be able to change the settings 2`] = `
"┏━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┓
┃     |  2  |  2  |  2  |  1  |  3  |  1  |  1  |  3  ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |  🥕 |     |     |     |     |     |  🥕 |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |     |     |  🥕 |  🥕 |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |     |     |     |     |  🥕 |  🥕 |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |     |     |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |  🥕 |  🥕 |     |     |     |  🥕 |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  3  |     |     |     |  🥕 |     |     |     |  🥕 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |     |  🥕 |  🥕 |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  3  |  🥕 |     |     |     |     |     |  🥕 |     ┃
┗━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┛"
`;

exports[`rabbitChase should be able to change the settings 3`] = `
"┏━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┓
┃     |  2  |  2  |  2  |  1  |  3  |  1  |  1  |  3  ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |  🥕 |     |     |     |  🐇 |     |  🥕 |  🐇 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |  🐇 |     |  🐇 |  🥕 |  🥕 |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |     |     |     |  🐇 |  🥕 |  🥕 |  🐇 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |  🐇 |     |     |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |  🥕 |  🥕 |     |     |     |  🥕 |  🐇 |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  3  |  🐇 |     |  🐇 |  🥕 |  🐇 |     |     |  🥕 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |     |  🥕 |  🥕 |     |     |  🐇 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  3  |  🥕 |  🐇 |     |  🐇 |     |  🐇 |  🥕 |     ┃
┗━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┛"
`;

exports[`rabbitChase should generate a game board 1`] = `
"┏━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┓
┃     |  1  |  1  |  2  |  0  |  3  |  0  ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |     |  🥕 |  🥕 |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  0  |     |     |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |  🥕 |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |  🥕 |     |     |  🥕 |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |     |     |     |  🥕 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |  🥕 |     |     |     |     ┃
┗━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┛"
`;

exports[`rabbitChase should generate a game board 2`] = `
"┏━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┓
┃     |  2  |  0  |  1  |  1  |  0  |  3  ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |  🥕 |     |     |     |     |  🥕 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |     |  🥕 |     |  🥕 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |     |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  0  |     |     |  🥕 |     |     |  🥕 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |  🥕 |     |     |     |     ┃
┗━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┛"
`;

exports[`rabbitChase should generate a game board 3`] = `
"┏━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┓
┃     |  2  |  0  |  2  |  1  |  0  |  2  ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  3  |     |     |     |  🥕 |  🥕 |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  0  |  🥕 |     |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |  🥕 |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  0  |     |     |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |  🥕 |     |     |  🥕 |     |  🥕 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |     |     |     |     |     ┃
┗━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┛"
`;

exports[`rabbitChase should render it to json correctly 1`] = `
Object {
  "carrots": Array [
    Object {
      "index": 0,
      "x": 0,
      "y": 0,
    },
    Object {
      "index": 4,
      "x": 4,
      "y": 0,
    },
    Object {
      "index": 9,
      "x": 3,
      "y": 1,
    },
    Object {
      "index": 15,
      "x": 3,
      "y": 2,
    },
    Object {
      "index": 25,
      "x": 1,
      "y": 4,
    },
    Object {
      "index": 28,
      "x": 4,
      "y": 4,
    },
    Object {
      "index": 32,
      "x": 2,
      "y": 5,
    },
  ],
  "cols": 6,
  "left": Array [
    3,
    0,
    1,
    1,
    0,
    2,
  ],
  "rows": 6,
  "top": Array [
    0,
    3,
    0,
    1,
    2,
    1,
  ],
}
`;

exports[`rabbitChase should render it to json correctly 2`] = `
"┏━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┓
┃     |  0  |  3  |  0  |  1  |  2  |  1  ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  3  |  🥕 |     |     |     |  🥕 |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  0  |     |     |     |  🥕 |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |     |  🥕 |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  0  |     |  🥕 |     |     |  🥕 |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |     |  🥕 |     |     |     ┃
┗━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┛"
`;

exports[`rabbitChase should render it to json correctly 3`] = `
"┏━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┯━━━━━┓
┃     |  0  |  3  |  0  |  1  |  2  |  1  ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  3  |  🥕 |  🐇 |     |  🐇 |  🥕 |  🐇 ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  0  |     |     |     |  🥕 |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |     |     |  🥕 |  🐇 |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  1  |     |  🐇 |     |     |     |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  0  |     |  🥕 |     |     |  🥕 |     ┃
┠─────┼─────┼─────┼─────┼─────┼─────┼─────┨
┃  2  |     |  🐇 |  🥕 |     |  🐇 |     ┃
┗━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┷━━━━━┛"
`;
